### PR TITLE
Provide support for In-memory template manager

### PR DESCRIPTION
**Pull request by JeethJJ**
Recreating PR : https://github.com/wso2-extensions/identity-event-handler-notification/pull/254

**Describe the issue:**
When introducing new notification templates (Email/SMS) and migrating customers to on-premises solutions, a data migration is currently required for existing tenants. However, with the removal of data migration support in upcoming Identity Server (IS) releases, we need to address this issue to ensure smooth template integration without relying on data migration.

**Proposed changes in this pull request**
Currently email/sms templates are stored in the registry, for any sub tenants we are not adding new templates when server startup. When creating any tenants we are adding default email templates to the registry per tenant, this leaves redundant data. With this fix, we are providing In-memory template manager which return default templates and act on read only mode.

Currently, it works with hybrid template manager and db-based template manager also.

Issue : https://github.com/wso2/product-is/issues/20996

### CODE REVIEW

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/DefaultTemplateManager.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        templatePersistenceManager.addNotificationTemplateType(displayName, notificationChannel, tenantDomain);
+    }
+
+    @Override
+    public boolean isNotificationTemplateTypeExists(String displayName, String notificationChannel, String tenantDomain)
+            throws NotificationTemplateManagerServerException {
+
+        return templatePersistenceManager.isNotificationTemplateTypeExists(displayName, notificationChannel,
+                tenantDomain) ||
+                inMemoryTemplateManager.isNotificationTemplateTypeExists(displayName, notificationChannel,
```
Comment (darshanasbg) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/DefaultTemplateManager.java:60]: Lets have inMemoryTemplateManager as the left operand to improve performance, since 'OR' operator's associativity is left to right. 

Basically this will allows to check existence for a template types, using in memory map first and then do a db call only when needed.

**Review by darshanasbg — CHANGES_REQUESTED** [review]

File: components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/OrganizationEmailTemplateTest.java [hunk 2]
```diff
@@ -406,6 +406,7 @@ private void mockRegistryResource(String notificationChannel, String displayName
 
         when(resourceMgtService.getIdentityResource(Matchers.anyString(), Matchers.anyString(), Matchers.anyString()))
                 .thenReturn(resource);
+        when(resourceMgtService.isResourceExists(Matchers.anyString(), Matchers.anyString())).thenReturn(true);
```
Comment (darshanasbg) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/OrganizationEmailTemplateTest.java:409]: Why we needed to mock this now?

Comment (JeethJJ) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/OrganizationEmailTemplateTest.java:409]: Due to the new if condition we have at [1]. Previously we did not check if it exists.
We need [2] to return true from the mocked registry resource. 

[1] https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/files#diff-55202b63368a9acaf49ac62f2c542327e0302c40fd60374d7378a6dcaaafd511R58-R60
[2] https://github.com/wso2-extensions/identity-event-handler-notification/blob/d610d0c389e5674ff60065a879f91d3835bb1210/components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/RegistryBasedTemplateManager.java#L91

Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/OrganizationEmailTemplateTest.java:409]: I'm still not clear, what the reason for this..

> Due to the new if condition we have at [1]. Previously we did not check if it exists.

You are implying the implementation of the `isNotificationTemplateTypeExists()` method was the root cause, but I think the root case don't have to do with the method implementation of `isNotificationTemplateTypeExists()` method.

I think, we have introduced a new code elsewhere, that caused for a new invocation of 'isNotificationTemplateTypeExists()` that wasn't there earlier..

By definition, if `isNotificationTemplateTypeExists()` method get called, it should have check for existence of the resource, so this mock should be there already..

Since we have to introduce this mock with this change means, we are now doing additional isExists() check in a flow elsewhere.. wanted to clear that this additional check is intentional.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/DefaultTemplateManager.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+                inMemoryTemplateManager.isNotificationTemplateExists(displayName, locale, notificationChannel,
+                        applicationUuid, tenantDomain);
+    }
+
+    @Override
+    public NotificationTemplate getNotificationTemplate(String displayName, String locale, String notificationChannel,
+                                                        String applicationUuid, String tenantDomain)
+            throws NotificationTemplateManagerServerException {
+
+        if (templatePersistenceManager.isNotificationTemplateExists(displayName, locale, notificationChannel,
```
Comment (darshanasbg) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/DefaultTemplateManager.java:109]: Lets check the possibility of removing this existence check in a get call.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class is responsible for storing notification templates from the email-admin-config.xml file.
+ * This class is primarily used to provide default notification templates without addition/modification,
+ * allowing for quick access of templates without migration.
+ * Templates are stored in nested maps, where the outer map is keyed by the template display name and the inner map
+ * is keyed by locale. This allows for easy retrieval of localized templates.
+ * This class used by {@link HybridTemplateManager} as a fallback mechanism.
```
Comment (darshanasbg) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:39]: HybridTemplateManager -> DefaultTemplateManager?

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+ * This class only supports for tenant specific notification templates.
```
Comment (darshanasbg) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:40]: `This class only supports for tenant specific notification templates.`
This comment is incorrect right?

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+
+        defaultEmailTemplates = populateTemplates(I18nMgtDataHolder.getInstance().getDefaultEmailTemplates());
+        defaultSMSTemplates = populateTemplates(I18nMgtDataHolder.getInstance().getDefaultSMSTemplates());
+    }
+
+    @Override
+    public void addNotificationTemplateType(String displayName, String notificationChannel, String tenantDomain)
+            throws NotificationTemplateManagerServerException {
+
+        // This method does nothing cause addition is not allowed for in-memory template manager.
```
Comment (darshanasbg) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:60]: better to throw not implemented exception?

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+     * @param notificationChannel The channel of the notification (e.g., Email, SMS).
+     * @param tenantDomain        The tenant domain.
+     * @return true if the template type exists, false otherwise.
+     * @throws NotificationTemplateManagerServerException if an error occurs during the operation.
+     */
+    @Override
+    public boolean isNotificationTemplateTypeExists(String displayName, String notificationChannel, String tenantDomain)
+            throws NotificationTemplateManagerServerException {
+
+        if (!StringUtils.isBlank(displayName)) {
```
Comment (darshanasbg) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:76]: Can use StringUtils.isNotEmpty()

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+            if (NotificationChannels.SMS_CHANNEL.getChannelType().equals(notificationChannel)) {
+                return defaultSMSTemplates.containsKey(displayName.toLowerCase());
```
Comment (darshanasbg) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:78]: displayName.toLowerCase()
better to extract to a variable, to avoid duplicated logic

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+     * @param tenantDomain        The tenant domain.
+     * @return true if the template type exists, false otherwise.
+     * @throws NotificationTemplateManagerServerException if an error occurs during the operation.
+     */
+    @Override
+    public boolean isNotificationTemplateTypeExists(String displayName, String notificationChannel, String tenantDomain)
+            throws NotificationTemplateManagerServerException {
+
+        if (!StringUtils.isBlank(displayName)) {
+            if (NotificationChannels.SMS_CHANNEL.getChannelType().equals(notificationChannel)) {
```
Comment (darshanasbg) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:77]: why the method `getTemplateMap()` not used here?

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+
+    @Override
+    public List<String> listNotificationTemplateTypes(String notificationChannel, String tenantDomain)
+            throws NotificationTemplateManagerServerException {
+
+        List<String> displayNames = new ArrayList<>();
+        Map<String, Map<String, NotificationTemplate>> defaultTemplates =  getTemplateMap(notificationChannel);
+        defaultTemplates.forEach((displayName, innerMap) ->
+                innerMap.forEach((locale, template) ->
+                        displayNames.add(template.getDisplayName())
```
Comment (darshanasbg) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:102]: If there are two templates for same type (like in two locale), aint this cased to return with a list that have  duplicate template types records?

File: components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManagerTest.java [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+        assertFalse(inMemoryBasedTemplateManager.isNotificationTemplateTypeExists(
+                negativeNotificationTemplate.getDisplayName(),
+                NotificationChannels.EMAIL_CHANNEL.getChannelType(), tenantDomain));
+        assertTrue(inMemoryBasedTemplateManager.isNotificationTemplateTypeExists(
+                positiveNotificationTemplate.getDisplayName(),
+                NotificationChannels.EMAIL_CHANNEL.getChannelType(), tenantDomain));
+    }
+
+    @Test
+    public void testListNotificationTemplateTypes() throws Exception {
```
Comment (darshanasbg) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManagerTest.java:103]: Add test/assert to verify the there are no duplicates records

**Review by darshanasbg — CHANGES_REQUESTED** [review]

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 10 above)
Comment (JeethJJ) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:102]: It does not make a difference, hence we return after removing the duplicates and merging [1].

[1] https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/files#diff-55202b63368a9acaf49ac62f2c542327e0302c40fd60374d7378a6dcaaafd511R200-R210

File: components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/OrganizationEmailTemplateTest.java (hunk 2 above)
Comment (JeethJJ) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/OrganizationEmailTemplateTest.java:409]: Previously provided point of reference was wrong. This is due to the new if condition we have at [1]

[1] https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/files#diff-55202b63368a9acaf49ac62f2c542327e0302c40fd60374d7378a6dcaaafd511R109

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 5 above)
Comment (JeethJJ) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:40]: It is correct, since we do not have an in-memory concern for application specific templates. They always can be retrieved from the DB

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 7 above)
Comment (JeethJJ) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:76]: `isNotEmpty` cant be used as it will allow empty strings with space. `isNotBlank` can be used.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 8 above)
Comment (JeethJJ) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:78]: This can be avoided when using getTemplateMap()

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 9 above)
Comment (JeethJJ) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:77]: Ack

File: components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManagerTest.java (hunk 11 above)
Comment (JeethJJ) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManagerTest.java:106]: ack

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 5 above)
Comment (Thumimku) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:41]: as @JeethJJ said with current logic we are not supporting application specific templates. But with some alterations we can support application specific templates. Since this application feature is introduced recently, we can improve the logic to send default template when we send GET request for application specific tempate.
@darshanasbg WDYT?

Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:41]: Aa ok.. Im not really referring to the application specific templates.

What I had in mind was `InMemoryBasedTemplateManager` not maintaining even tenant **specific** templates.. `InMemoryBasedTemplateManager` is keeping set of templates that will be commonly used as default templates across all the tenants.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/OrganizationEmailTemplateTest.java (hunk 2 above)
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/OrganizationEmailTemplateTest.java:409]: We have removed this if condition with https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/commits/8628c42d5f74b1e58a734be7fc641c3f846af46d.
So shall we remove this additional mock?

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/DefaultTemplateManager.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    @Override
+    public boolean isNotificationTemplateExists(String displayName, String locale, String notificationChannel,
+                                                String applicationUuid, String tenantDomain)
+            throws NotificationTemplateManagerServerException {
+
+        return templatePersistenceManager.isNotificationTemplateExists(displayName, locale, notificationChannel,
+                applicationUuid, tenantDomain) ||
+                inMemoryTemplateManager.isNotificationTemplateExists(displayName, locale, notificationChannel,
```
[the remark this answers is not in the document]
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/DefaultTemplateManager.java]: Fixed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/commits/8628c42d5f74b1e58a734be7fc641c3f846af46d

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/DefaultTemplateManager.java (hunk 1 above)
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/DefaultTemplateManager.java]: Fixed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/commits/8628c42d5f74b1e58a734be7fc641c3f846af46d

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 4 above)
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java]: Fixed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/commits/8628c42d5f74b1e58a734be7fc641c3f846af46d

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 6 above)
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java]: Fixed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/commits/8628c42d5f74b1e58a734be7fc641c3f846af46d

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 7 above)
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java]: Fixed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/commits/8628c42d5f74b1e58a734be7fc641c3f846af46d

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 8 above)
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java]: Fixed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/commits/8628c42d5f74b1e58a734be7fc641c3f846af46d

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 9 above)
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java]: Fixed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/commits/8628c42d5f74b1e58a734be7fc641c3f846af46d

File: components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManagerTest.java (hunk 11 above)
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManagerTest.java:106]: Fixed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/commits/8628c42d5f74b1e58a734be7fc641c3f846af46d

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/DefaultTemplateManager.java (hunk 3 above)
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/DefaultTemplateManager.java]: Fixed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/commits/8628c42d5f74b1e58a734be7fc641c3f846af46d

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 5 above)
Comment (Thumimku) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:41]: I revisited the existing code 
ref: https://github.com/wso2-extensions/identity-event-handler-notification/blob/master/components/event-handler-notification/org.wso2.carbon.identity.event.handler.notification/src/main/java/org/wso2/carbon/identity/event/handler/notification/util/NotificationUtil.java#L724

Here when there is no application template found, it fall backs to organisation wise template, to support this behaviour we have to keep the logic as it is.  Other wise, customer has to configure email template for each and every application. Hence +1 to have this logic to skip when application id presented.

Comment (JeethJJ) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:41]: Ack

**Comment by darshanasbg** [comment]
Also with this change we dont need to persist default templates with server startup[1] and tenant creation[2].. So lets remove those and associate code that become unused with that.. 
In fact we can remove,
* whole TenantManagementListener.
* Remove `loadDefaultEmailTemplates()` `loadDefaultSMSTemplates()` method impls
* EmailTemplateManager.addDefaultEmailTemplates() method deprecated
* EmailTemplateManagerImpl.addDefaultEmailTemplates() method deprecated & clean the impl to throw not implemented exception.
* NotificationTemplateManager.addDefaultNotificationTemplates() method deprecated
* EmailTemplateManagerImpl.addDefaultNotificationTemplates() method deprecated & clean the impl to throw not implemented exception.

[1] https://github.com/wso2-extensions/identity-event-handler-notification/blob/d610d0c389e5674ff60065a879f91d3835bb1210/components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java#L144-L146
[2] https://github.com/wso2-extensions/identity-event-handler-notification/blob/d610d0c389e5674ff60065a879f91d3835bb1210/components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/TenantManagementListener.java#L48-L56

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 5 above)
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java:41]: @Thumimku Im not exactly clear on the change you are proposing.

Note that my intention is only to correct the quoted the class comment to be more accurate, not to do any logic change of the code execution.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/OrganizationEmailTemplateTest.java (hunk 2 above)
Comment (JeethJJ) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/test/java/org/wso2/carbon/email/mgt/OrganizationEmailTemplateTest.java]: Ack, yes. Removed.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 5 above)
Comment (JeethJJ) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java]: Updated the comment.

Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java]: To be more direct, the class comment can be reworded as,

/**
 * This class is responsible for handling default templates provided by the system.
 * 
 * System default templates are used as default org notification templates in a given tenant and based on the content of the email-admin-config.xml.
 * Any modifications to system default templates needs to be done via email-admin-config.xml and any such changes will be reflected only with the server startup.
 * 
 * This class have support for get, list, exists operations and any invocations to modification operations throws NotImplementedException.
 * 
 * This class expected to be use only with conjunction with another {@link TemplatePersistenceManager} implementation which supports full CRUD operations, hence {@link DefaultTemplateManager} provides that aggregation using this as a fallback provider.
 */

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 6 above)
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java]: Better to throw UnsupportedOperationException as its more accurate one.
Ref: https://stackoverflow.com/a/22045192/565835

**Comment by JeethJJ** [comment]
Changes suggested through [1] will be addressed in another PR.

[1] https://github.com/wso2-extensions/identity-event-handler-notification/pull/256#issuecomment-2316811960

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java (hunk 5 above)
Comment (JeethJJ) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/InMemoryBasedTemplateManager.java]: Updated with https://github.com/wso2-extensions/identity-event-handler-notification/pull/256/commits/5d985d4254d5681ad87d157dc8a8a8c37cd69ff9

**Review by darshanasbg — DISMISSED** [review]

**Review by darshanasbg — APPROVED** [review]
